### PR TITLE
Add redux-persist

### PR DESCRIPTION
**Pull request by tifa2UP**
### Summary
Added [redux-persist](https://github.com/rt2zz/redux-persist) to persist parts of our redux store in localstorage across sessions. This will be useful for adding features such as showing the last read position.

Followed the set-up guide here: https://redux-toolkit.js.org/usage/usage-guide#use-with-redux-persist.

Will create a follow-up PR insha'Allah to demo the values being saved in localstorage and persisting across sessions. 

### Test Plan
Confirmed that a `persist:root` object is being created in localstorage (see screenshot). No observable changes in the app behavior. Tested the usage and confirmed that the whitelisted values persist in #71. 

### Screenshots

image

### CODE REVIEW

File: src/pages_/_app.tsx [hunk 1]
```diff
@@ -2,19 +2,22 @@ import React from 'react';
...
 
 function MyApp({ Component, pageProps }): JSX.Element {
   return (
     <Provider store={store}>
-      <ThemeProvider theme={theme}>
-        <DefaultSeo {...createSEOConfig({})} />
-        <Component {...pageProps} />
-        <Footer />
-      </ThemeProvider>
+      <PersistGate loading={null} persistor={persistor}>
```
Comment (tifa2UP) [inline src/pages_/_app.tsx:14]: The loading time is too small imo to add a loading component, it's helpful for apps that store their full redux store in localstorage. We can reassess and change it as our app gets bigger.

Comment (nabeel-) [reply src/pages_/_app.tsx:14]: Is the loading time negligible across all devices? If there are some low-powered devices that might experience a noticeable delay we could get the best of both worlds by having a “delayed spinner” that will only kick in after a timeout. In that case, we avoid flashing a loading spinner on devices where the delay is minimal and show indications to the user that something is happening on lower-powered devices that take a little longer.

Not something to block this particular PR on - just a drive-by thought!

Comment (tifa2UP) [reply src/pages_/_app.tsx:14]: 1) Yeah, the loading time is niglible on all devices, if there's no localstorage data, the `PersistGate` component loads the application right away. So there's no observable difference for first-time users and web crawlers. For returning users, the storage payload is pretty small, and the data is read locally, I'm suspecting that it'll be neglible too. 

2) I'm open to the idea of having a loading component that appears after 500ms or so, we should do a performance rundown once we have a working version of the app iA and identify the bottlenecks :)

Comment (mmahalwy) [reply src/pages_/_app.tsx:14]: Hmm. after looking at the code, this will show empty on SSR, no?

Comment (tifa2UP) [reply src/pages_/_app.tsx]: Good catch @mmahalwy. Created `Provider.tsx` to fix this issue. 

We now only load `<PersistGate />` on the client. I confirmed that the DOM is loaded for SSG, and that redux-persist still works as expected

**Comment by AhmedCodeGuy** [comment]
Looks good to me!

**Comment by tifa2UP** [comment]
> With changes to src/redux/Provider.tsx, this would still cause a diff in server rendering and client mounting. Might be some inconsistency here

This seems to be recommended approach to use redux-persist with SSR. Do you have concerns for not rendering `<PeristGate />` on the server?

**Comment by mmahalwy** [comment]
> > With changes to src/redux/Provider.tsx, this would still cause a diff in server rendering and client mounting. Might be some inconsistency here
> 
> This seems to be recommended approach to use redux-persist with SSR. Do you have concerns for not rendering `<PeristGate />` on the server?

I feel like this probably would have an inconsistent flash. Let's try it though?

**Pull request merged**